File: Beecrowd_Parte5/Turma_Dividida_3484.java
import java.util.*;

class TreeNode {
    int height;
    TreeNode left;
    TreeNode right;

    TreeNode(int height) {
        this.height = height;
        this.left = null;
        this.right = null;
    }
}

public class Turma_Dividida_3484 {
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        // Leitura do número de alunos
        int n = scanner.nextInt();
        int[] heights = new int[n];
        
        // Leitura das alturas dos alunos
        for (int i = 0; i < n; i++) {
            heights[i] = scanner.nextInt();
        }
        
        // Construção da árvore
        TreeNode root = new TreeNode(heights[0]);
        for (int i = 1; i < n; i++) {
            insertNode(root, heights[i]);
        }
        
        // Obtenção da menor altura em cada nível
        List<Integer> result = findMinHeightsByLevel(root);
        
        // Impressão dos resultados
        for (int i = 0; i < result.size(); i++) {
            System.out.println(i + " " + result.get(i));
        }
        
        scanner.close();
    }
    
    // Função para inserir um nó na árvore
    public static void insertNode(TreeNode root, int height) {
        TreeNode current = root;
        while (true) {
            if (height < current.height) {
                if (current.left == null) {
                    current.left = new TreeNode(height);
                    break;
                } else {
                    current = current.left;
                }
            } else {
                if (current.right == null) {
                    current.right = new TreeNode(height);
                    break;
                } else {
                    current = current.right;
                }
            }
        }
    }
    
    // Função para encontrar a menor altura em cada nível usando BFS
    public static List<Integer> findMinHeightsByLevel(TreeNode root) {
        List<Integer> minHeights = new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        
        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            int minHeight = Integer.MAX_VALUE;
            
            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.poll();
                minHeight = Math.min(minHeight, node.height);
                
                if (node.left != null) {
                    queue.add(node.left);
                }
                if (node.right != null) {
                    queue.add(node.right);
                }
            }
            
            minHeights.add(minHeight);
        }
        
        return minHeights;
    }
}

File: Beecrowd_Parte5/Turma_Dividida_3484.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

typedef struct TreeNode {
    int height;
    struct TreeNode *left, *right;
} TreeNode;

TreeNode* createNode(int height) {
    TreeNode* newNode = (TreeNode*)malloc(sizeof(TreeNode));
    newNode->height = height;
    newNode->left = newNode->right = NULL;
    return newNode;
}

TreeNode* insertNode(TreeNode* root, int height) {
    if (root == NULL) {
        return createNode(height);
    }

    if (height < root->height) {
        root->left = insertNode(root->left, height);
    } else {
        root->right = insertNode(root->right, height);
    }
    return root;
}

void findMinHeightsByLevel(TreeNode* root) {
    if (root == NULL) return;
    
    int level = 0;
    int levelSize = 1;
    int nextLevelSize = 0;
    TreeNode* queue[250];
    int front = 0, rear = 0;
    int minHeight = INT_MAX;
    
    queue[rear++] = root;
    
    while (front < rear) {
        TreeNode* node = queue[front++];
        if (node->height < minHeight) {
            minHeight = node->height;
        }
        
        if (node->left != NULL) {
            queue[rear++] = node->left;
            nextLevelSize++;
        }
        if (node->right != NULL) {
            queue[rear++] = node->right;
            nextLevelSize++;
        }
        
        if (--levelSize == 0) {
            printf("%d %d\n", level, minHeight);
            level++;
            levelSize = nextLevelSize;
            nextLevelSize = 0;
            minHeight = INT_MAX;
        }
    }
}

int main() {
    int n;
    scanf("%d", &n);

    int heights[n];
    for (int i = 0; i < n; i++) {
        scanf("%d", &heights[i]);
    }

    TreeNode* root = NULL;
    for (int i = 0; i < n; i++) {
        root = insertNode(root, heights[i]);
    }

    findMinHeightsByLevel(root);

    return 0;
}
